base/server/src/main/java/com/netscape/cms/password/PasswordChecker.java

```diff
@@ -27,14 +35,124 @@
+    private int substringMatch = 0;
+    private int maxRepeatedChar = 0;
+    private boolean cracklibCheck = false;
     /**
      * Default constructor.
      */
     public PasswordChecker() {
     }
 
+    public PasswordChecker(int minSize, int minUpperLetter, int minLowerLetter, int minNumber, int minPunctuationChar,
+            int substringMatch, int maxRepeatedChar, boolean cracklibCheck) {
+        this.minSize = minSize;
+        this.minUpperLetter = minUpperLetter;
+        this.minLowerLetter = minLowerLetter;
+        this.minNumber = minNumber;
+        this.minPunctuationChar = minPunctuationChar;
+        this.substringMatch = substringMatch;
+        this.maxRepeatedChar = maxRepeatedChar;
+        this.cracklibCheck = cracklibCheck;
+    }
+
+    public PasswordQuality checkPassword(String password) throws EPasswordCheckException{
+        if (password == null || password.length() == 0) {
+            return PasswordQuality.CMS_PASSWORD_EMPTY_PASSWORD;
+        }
+
+        if (password.length() < minSize) {
+            return PasswordQuality.CMS_PASSWORD_INVALID_LEN;
+        }
+
+        String basePassword = null;
+        if (minUpperLetter + minLowerLetter + minNumber + minPunctuationChar > 0) {
+            StringBuilder base = new StringBuilder(password.replaceAll("^\\p{Upper}*", ""));
+            base.reverse();
+            basePassword = base.toString().replaceAll("^\\d*", "");
+        }
+        if (minUpperLetter > 0 && countLetters(basePassword, "\\p{Upper}") < minUpperLetter) {
+            return PasswordQuality.CMS_PASSWORD_MISSING_UPPER_CASE;
+        }
+        if (minLowerLetter > 0 && countLetters(basePassword, "\\p{Lower}") < minLowerLetter) {
+            return PasswordQuality.CMS_PASSWORD_MISSING_LOWER_CASE;
+        }
+        if (minNumber > 0 && countLetters(basePassword, "\\d") < minNumber) {
+            return PasswordQuality.CMS_PASSWORD_MISSING_NUMERIC;
+        }
+        if (minPunctuationChar > 0 && countLetters(basePassword, "\\p{Punct}") < minPunctuationChar) {
+            return PasswordQuality.CMS_PASSWORD_MISSING_PUNCTUATION;
+        }
+
+
+        if (substringMatch > 0) {
+            for (int i = 0; i< password.length() - (substringMatch * 2); i++) {
+                String seq = password.substring(i, i + substringMatch);
+                String invSeq = new StringBuilder(seq).reverse().toString();
+                if(password.indexOf(seq, i + substringMatch) > 0 || password.indexOf(invSeq, i + substringMatch) > 0){
+                    return PasswordQuality.CMS_PASSWORD_SEQUENCE;
+                }
+            }
+        }
+
+        if (maxRepeatedChar > 0) {
+            int mostRepeatedChar = password.chars().mapToObj(x -> (char) x)
+                    .collect(Collectors.groupingBy(x -> x, Collectors.counting()))
+                    .entrySet().stream().max(Map.Entry.comparingByValue())
+                    .get().getValue().intValue();
```

Comment:
If I'm not mistaken, if this is the attempt to catch " repeated characters", I suspect that's misunderstood.  I think "repeated characters" means something like "aaa", "bb" etc. instead of, say, "takuaba", where 'a' appears 3 times but it's not a meaningful word and doesn't have consecutive repeated characters so should be fine.

Reply:
In fact, I think the previous check (substringMatch) should catch repeated characters already, right?  If so, maybe remove this check.  I think we can call this "Phase 1" and add improvements later.

Reply:
I do not think the substringMatch capture repeated characters. If a password contains `aaaaabbbbb` and we check with substring of 6 then it is OK but we could have max repeated of 3 which should fails. I'll update the check.

Reply:
yeah, maybe you can change the logic so that the value assigned to substringMatch means that it will check with substringMatch,  substringMatch-1, and  substringMatch-2, etc.